perf: BOM Update Tool

marination opened the pull request:
<details><summary> First Draft </summary>

##### `get_boms_in_bottom_up_order`
- Create child-parent map once and fetch value from child key to get parents
- Get parents recursively for a leaf node (get all ancestors)
- Approx. 44 secs for 4lakh 70k BOMS (used to go on for over 5-10 mins earlier, maybe more idk)

##### `update_cost`
- Remove double RM cost calculation
- Update child items in db as well (optionally)
- Don't regenerate exploded items for each BOM while updating cost (only in update cost job/bom function)🤦🏻‍♀️
- Approx 5 minutes for 20k BOMs. [WIP]
</details>

> There were correctness issues that could all be solved with a newer AlGoRiThM, that took preference. So first correctness >> then perf

### New BOM Cost Updation Flow
> **_Dependants_** = Immediate Parents, **_Dependencies_** = Children (parent-child is not accurate really)

Consider the following BOMs and their dependencies:

![BOM-structure drawio](https://user-images.githubusercontent.com/25857446/170047042-e85b90e1-5bdf-4d5b-926e-5bde51542e46.png)
- While updating all BOMs, we update the **Raw Material BOMs** or leaf nodes first and work our way up
- An important thing here: **We must only update a BOM after all its _Dependencies_ are resolved/updated**
- That means: α (alpha) cannot be updated until A,B,1,2,3,4 are updated
- For this we update BOMs **one level at a time**, and **within a level we process BOMs parallelly as they are independent.**
- BOMs **within each level** are split into batches of 20k (takes 5 mins) and queued for updation
- Once a level is done a Cron Job will prep and start processing the next level

#### Working:
- Given the above example, on submitting the BOM Update Log, **Raw Material BOMs** will be processed first (batch size = 20k). The status will be **In Progress**

  <img src="https://user-images.githubusercontent.com/25857446/170224003-1d87b8cd-9264-4b02-8da0-30e37255465b.png" width="60%">
- These Jobs will keep updating their row in the table in the Log and keep checking if all other jobs of this level are done (via `bom_batches` table)
- The **Current Level** is also maintained in the Progress Section, which will increment as we progress up the levels
- **Once a level is complete** and the cron (every 5 mins) determines a level is complete (all rows of current level have status **Pending** -> **Completed**), it will prepare next level parent boms (dependants) and kickstart the next level

   <img width="70%" alt="Screenshot 2022-05-24 at 2 17 49 PM" src="https://user-images.githubusercontent.com/25857446/172573240-12d10bb3-c910-4dd0-bfb4-26facf74d714.png">
- **How do we determine "Parent BOMs" ?** : By checking immediate parents with no **_Dependencies_**. Here **1** has immediate parents **A** and **α (alpha)**. **A** can be considered since it has no unresolved/non-updated **_Dependencies_**, but **α (alpha)** has  A and B as unresolved dependencies. **So we proceed with A as parent of 1 in the next level**.
- A Cron Job will run every 5 minutes and check for **In Progress** jobs and start processing the next level **if all jobs of current level are completed**
- The Log will be set as **Complete** when there are no more **_Dependants_/ Parent BOMs** to resolve/update

-----------------------------
 > #### Important change/fix to test
> - Rate updation as per **Raw Material Cost as per** used to not work, which now does
> - Rates are reset rightfully on save and fetched as per Valuation/Last Purchase/Price List
----------------------------
**To Test:**
- Create BOM with 3 levels (RM BOM, Subassembly BOM that uses RM BOM, FG BOM that uses Subassemble BOM)
- Set RM cost as per **Valuation Rate** > Change valuation rate of Item in RM BOM > Update all costs via tool > check
- Set RM cost as per **Valuation Rate** with different BOM currency (USD)> Change valuation rate of Item in RM BOM > Update all costs via tool > check
- Set RM cost as per **Price List** > Change Price List rate of Item in RM BOM > Update all costs via tool > check
- Set RM cost as per **Price List** in different Price List currency > Change Price List rate of Item in RM BOM > Update all costs via tool > check

**Todo:**

- [x] Clearing BOM Update Logs periodically (or atleast the table in it)

marination committed:
perf: `get_boms_in_bottom_up_order`

- Create child-parent map once and fetch value from child key to get parents
- Get parents recursively for a leaf node (get all ancestors)
- Approx. 44 secs for 4lakh 70k boms

marination committed:
fix: DB update child items, remove redundancy, fix perf

- Move `get_boms_in_bottom_up_order` in bom update tool’s file
- Remove repeated rm cost update from `update_cost`. `calculate_cost` handles RM cost update
- db_update children in `calculate_cost` optionally
- Don’t call `update_exploded_items` and regenerate exploded items in `update_cost`. They will stay the same (except cost)

marination committed:
fix: Call `calculate_cost` for Draft BOM and typo in argument

marination committed:
perf: Use cached doc instead of `get_doc`

- Doc is only used to iterate over items(which wont change) and change rate/amount of rows
- These changes are inserted in db via `db_update`, so no harm
- Tested locally: refetching cached doc after db update, reflects fresh data.

marination committed:
fix: Get fresh RM rate in `calculate_rm_cost`

marination committed:
fix: `test_work_order_with_non_stock_item`

- Use the right price list and currency to avoid rate conversion (1000/62.9), since rates are reset correctly now
- Use RM rate based on Price List in BOM. Non stock item has no valuation

marination committed:
feat: Level-wise BOM cost updation

- Process BOMs level wise and Pause after level is complete
- Cron job will resume Paused jobs, which will again process the new level and pause at the end
- This will go on until all BOMs are updated
- Added Progress section with fields to track updated BOMs in Log
- Cleanup: Add BOM Updation utils file to contain helper functions/sub-functions
- Cleanup: BOM Update Log file will only contain functions that are in direct context of the Log

Co-authored-by: Gavin D'souza <gavin18d@gmail.com>

marination committed:
style: Update docstrings and fix/add type hints + Collapsible progress section in Log

marination commented:
Update cost tests are breaking. Will fix while writing tests

marination committed:
feat: Only update exploded items rate and amount

- Generate RM-Rate map from Items table (will include subassembly items with rate)
- Function to reset exploded item rate from above map
- `db_update` exploded item rate only if rate is changed
- Via Update Cost, only update exploded items rate, do not regenerate table again
- Exploded Items are regenerated on Save and Replace BOM job
- `calculate_exploded_cost` is run only via non doc events (Update Cost button, Update BOMs Cost Job)

marination committed:
chore: Change BOM Progress field types to Long Text

marination committed:
perf: `get_next_higher_level_boms`

- Separate getting dependants and checking if they are valid (loop within loop led to redundant processing that slowed down function)
- Adding to above, the same dependant(parent) was repeatedly processed as many children shared it. Expensive.
- Use a parent-child map similar to child-parent map to check if all children are resolved
- `map.get()` reduced time: 10 mins -> 0.9s~1 second (as compared to `get_cached_doc` or query)
- Total time: 17 seconds to process 6599 leaf boms and 4.2L parent boms
- Previous Total time: >10 mins (I terminated it due to not wanting to waste time XD)

marination committed:
fix: Safe cast `row.rate` (in case of faulty exploded items, edge case but oh well)

marination committed:
chore: Limit Update Cost jobs & `db_update` only if changed values

- If `Update Cost` job is ongoing, then block creation of new ones since all BOMs are updated
- `db_update` in `calculate_rm_cost` only if changed values to reduce redundant row updates
- Misc: Use variable for batch size

marination committed:
feat: Track progress in Log Batch/Job wise

- This was done due to stale reads while the background jobs tried updating status of the log
- Added a table where all bom jobs within log will be tracked with what level they are processing
- Cron job will check if table jobs are all processed every 5 mins
- If yes, it will prepare parents and call `process_boms_cost_level_wise` to start next level
- If pending jobs, do nothing
- Current BOM Level is being tracked that helps adding rows to the table
- Individual bom cost jobs (that are queued) will process and update boms > will update BOM Update Batch table row with list of updated BOMs

marination committed:
chore: Miscellanous fixes/enhancements

- `get_valuation_rate`: if no bins are found return 0, SLEs do not exist either
- `get_valuation_rate`: Compute average valuation rate via query
- `get_rm_rate_map`: set order_by as None to avoid creating sort index (modified) each time query runs (seen in process list)
- BOM Update Batch: add status field and hide `boms_updated` so that  users can see progress without loading all updated boms (too much data)
- BOM Update Batch: set batch row status to completed after job runs
- BOM Update Log: remove `parent_boms` field (just pass parent boms to processing function) & remove Paused state (not used)
- Move job to long queue to avoid choking default queue
- `update_cost_in_boms`: use `get_doc` as each BOM is accessed only once. Use `for_update` to lock BOM row
- Commit after every 100 BOMs

marination committed:
chore: `get_valuation_rate` in bom.py must always return float & goto Item master if no bins

marination commented:
Was an issue with concurrent jobs and incorrect locking sequence causing lost connections to db server.
Fixed and works as expected.
4 jobs parallelly take ~ 12-14 minutes (4*20k = 80,000 BOMs)

marination committed:
test: Util to update cost in all BOMs

- Utility to update cost in all BOMs without cron jobs or background jobs (run immediately)
- Re-use util wherever all bom costs are to be updated
- Skip explicit commits if in test
- Specify company in test records (dirty data sometimes, company wh mismatch)
- Skip background jobs queueing if in test

marination committed:
Merge branch 'develop' into perf-bom-update-tool

marination committed:
test: Fix `test_update_bom_cost_in_all_boms`

- Use base_rate for assertions as rate is subject to change due to conversion factor (USD)

marination committed:
chore: `get_valuation_rate` sider fixes

- Use qb instead of db.sql
- Don't use `args` as argument for function
- Cleaner variable names

ankush commented on erpnext/manufacturing/doctype/bom_update_log/bom_update_log.py:
```diff
@@ -42,123 +51,192 @@ def validate_bom_items(self):
-def get_parent_boms(new_bom: str, bom_list: Optional[List] = None) -> List:
-	bom_list = bom_list or []
-	bom_item = frappe.qb.DocType("BOM Item")
+		frappe.db.auto_commit_on_many_writes = 1
```
This can be removed now (there are commits at N processed BOMs)

marination replied:
I think you meant to remove it in `update_cost_in_level` 😄

ankush replied:
Yep, this one can stay 😬

ankush commented on erpnext/manufacturing/doctype/bom_update_log/bom_update_log.py:
```diff
@@ -42,123 +51,192 @@ def validate_bom_items(self):
-		frappe.qb.from_(bom).select(bom.total_cost / bom.quantity).where(bom.name == new_bom).run()
-	)
+	if frappe.flags.in_test:
+		return current_boms, current_level
```
Keeping prod and test code similar will help. 

We can use `enqueue(..., now=frappe.flags.in_test)` to run things in tests immediately. Doesn't simulate IRL behavior but "close enough"

ankush commented on erpnext/manufacturing/doctype/bom_update_log/bom_update_log.py:
```diff
@@ -42,123 +51,192 @@ def validate_bom_items(self):
-		doc.db_set("status", "Completed")
+		if frappe.flags.in_test:
+			# skip background jobs in test
+			return boms_to_process, batch_row.name
```
Same comment as before, most flags.in_test conditions (except commits) seem to/should be replacable.

ankush commented on erpnext/manufacturing/doctype/bom_update_log/test_bom_update_log.py:
```diff
@@ -55,42 +63,59 @@ def test_bom_update_log_validate(self):
-		log2 = enqueue_replace_bom(
-			boms=self.boms,
+
+def update_cost_in_all_boms_in_test():
```
This *might* be replaceable with actual prod code using `now`. That will be more maintainable in case design/logic changes in future.

ankush commented on erpnext/manufacturing/doctype/bom_update_log/bom_updation_utils.py:
```diff
@@ -0,0 +1,224 @@
+		bom_obj.calculate_cost()
+		bom_obj.update_parent_cost()
+		bom_obj.db_update()
+		if bom_obj.meta.get("track_changes") and not bom_obj.flags.ignore_version:
+			bom_obj.save_version()
```
```suggestion
		bom_obj.save_version()
```

both condition already seem to be part of `Document.save_version`


Also check if version is getting created, since `_doc_before_save` is required which isn't triggered unless post save methods are run... maybe it can be added manually. 


PS: tested locally, no version docs found for bom.

ankush replied:
On second thought generating versions for all BOMs, every day seems a bit too much. 😅

marination replied:
fixed locally, version was not created because the faux `_doc_before_save` pointed to the bom_obj so there was never any difference between the two. Just had to make a copy and assign to `_doc_before_save`

ankush commented on erpnext/manufacturing/doctype/bom_update_log/bom_update_log.py:
```diff
@@ -42,123 +51,192 @@ def validate_bom_items(self):
+		current_level = 0
+		current_boms = get_leaf_boms()
+		values = {
+			"processed_boms": json.dumps({}),
```
This can grow quite large in DB for numbers we are dealing with. Maybe add a weekly/monthly job to discard old successful BOM update logs. 


reminds me of this: https://github.com/frappe/frappe/issues/17014

marination replied:
I can clear this field once the log is complete. Its kinda useless after completion. Its used to mainly check if lower levels were processed while going to next level.

> Maybe add a weekly/monthly job to discard old successful BOM update logs.

How old is old 🙈 ? @ankush

ankush replied:
Most logs are deleted at 30 days intervals in log settings... lemme see if we can speed up doctype agnostic log settings 👀

ankush replied:
Just clearing that field might be cool too 👍

gavindsouza replied:
Just emptying the child table (`bom_batches` ) upon update success should do. Let's not add yet another cleanup job for failed tasks? That bit can be manual - or handled by the likes of "Log Settings".

marination replied:
@ankush will work on auto clearing logs in a separate PR. Need some clarity on what data should be maintained from user POV. For now `processed_boms` is cleared

ankush committed:
Merge branch 'develop' into perf-bom-update-tool

ankush commented:
Tested with deeply nested BOM with different costs for each RM and some ops cost. 

#### RM cost
- Started with everything valued @ 0
- Update all bottom-most item (RM) to some valuation rate and did BOM update (scheduled job trigger manually)

Exploded item cost == individual subassy costs. 
<img width="1424" alt="Screenshot 2022-06-09 at 2 05 45 PM" src="https://user-images.githubusercontent.com/9079960/172803120-c3fc9e36-4336-4065-bb62-f7703818a0ca.png">

---


#### Ops cost

- Used same BOM, revised and replaced a revision with ops. 
- Ran update tool
<img width="1290" alt="Screenshot 2022-06-09 at 2 20 14 PM" src="https://user-images.githubusercontent.com/9079960/172806255-93f1be4f-6ca0-4d2a-93b0-678dabe30346.png">


Increased levels  (by moving this full subassy on another FG)

<img width="1067" alt="Screenshot 2022-06-09 at 2 29 02 PM" src="https://user-images.githubusercontent.com/9079960/172808323-86e6b2b2-abcf-4bdd-adb6-81a1253868d9.png">



Partitioning / update logic LGTM 👍 


---

unrelated feature request (I added these columns in BOM explorer (v. hacky implementation) but this might help in actually exploring costs 😬 )

marination committed:
chore: Less hacky tests, versioning (replace bom) and clearing log data (update cost)

- Remove `auto_commit_on_many_writes` in `update_cost_in_level()` as commits happen every N BOMs
- Auto commit every 50 BOMs
- test: Remove hacky `frappe.flags.in_test` returns
- test: Enqueue `now` if in tests (for update cost and replace bom)
- Replace BOM: Copy bom object to `_doc_before_save` so that version.py finds a difference between the two
- Replace BOM: Add reference to version
- Update Cost: Unset `processed_boms` if Log is completed (useless after completion)
- test: `update_cost_in_all_boms_in_test` works close to actual prod implementation (only call Cron job manually)
- Test: use `enqueue_replace_bom`  so that test works closest to production behaviour

Co-authored-by: Ankush Menat <ankushmenat@gmail.com>
